### PR TITLE
[feat] Monday Adoptee Status Update Function

### PR DESCRIPTION
**Pull request by niranjana2222**
[//]: # "Feel free to customize this template and the emojis to your project's vibes"

### Description

Adds a Monday.com helper in `actions/monday/mutations/changeStatus.ts` to update Adoptee status. 

- if status == 'OFC': no change is made
- if status == 'WL' reads from Supabase to map the id to 'WLFA' or 'WL'

Replaces the old helper in `exportApplication.ts` with `updateAdopteeMondayStatus`

## How to review
[//]: # 'Required - Describe the order in which to review files and what to expect when testing locally. Is there anything specifically you want feedback on? Should this be reviewed commit by commit, or all at once? What are some user flows to test? What are some edge cases to look out for?'

1. Look at `actions/monday/mutations/changeStatus.ts` for mutation behavior, Supabase query, and error handling.
2. Check `actions/monday/mutations/exportApplication.ts` for updated call.

### CODE REVIEW

File: actions/monday/mutations/changeStatus.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+  const uniqueAdopteeMondayIds = Array.from(adopteeMondayIds);
+
+  const statusLabelsById =
+    status === 'WL'
+      ? await getWaitListStatusLabels(uniqueAdopteeMondayIds)
+      : Object.fromEntries(
+          //for not string string
+          uniqueAdopteeMondayIds.map(id => [id, OFC_STATUS_LABEL]),
+        );
```
Comment (jinkang-0) [inline actions/monday/mutations/changeStatus.ts]: `getWaitListStatusLabels` can throw an error, this should be properly handled so the server doesn't crash

File: actions/monday/mutations/changeStatus.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+//main function
+export async function updateAdopteeMondayStatus(
+  adopteeMondayIds: string[],
+  status: MondayAdopteeStatus,
+) {
+  if (adopteeMondayIds.length === 0) return '';
+
+  const uniqueAdopteeMondayIds = Array.from(adopteeMondayIds);
```
Comment (jinkang-0) [inline actions/monday/mutations/changeStatus.ts]: can I ask why you wrote this? afaik this simply copies the array, which shouldn't be necessary for the operations we're doing on `adopteeMondayIds`. The name is also a misnomer, it's not making it unique in any way.

File: actions/monday/mutations/changeStatus.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+      }
+    `;
+
+    try {
+      await mondayApiClient.request(mutationQuery);
+    } catch (error) {
+      Logger.error(
+        `Failed to update adoptee Monday status to WL for ids ${uniqueAdopteeMondayIds.join(',')}: ${error}`,
+      );
+      throw error;
```
Comment (jinkang-0) [inline actions/monday/mutations/changeStatus.ts]: server actions shouldn't throw an error. This not only crashes the server, it provides no meaningful message to the client that issued the request. Instead, opt for a `{ data, error }` return object paradigm. Just make sure this is consistent even in the case of successful calls, e.g. `{ data, error: null }`

**Review by jinkang-0 — COMMENTED** [review]
great work niranjana! left some comments for error handling, but otherwise looks good

File: actions/monday/mutations/changeStatus.ts (hunk 1 above)
Comment (niranjana2222) [reply actions/monday/mutations/changeStatus.ts]: oh okay, I'll fix that, but also, how would I know that it would throw an error?

File: actions/monday/mutations/changeStatus.ts (hunk 2 above)
Comment (niranjana2222) [reply actions/monday/mutations/changeStatus.ts]: Yes, I'll fix that, I thought I needed to make it an Array if it wasn't one but realized it's unnecessary

File: actions/monday/mutations/changeStatus.ts (hunk 3 above)
Comment (niranjana2222) [reply actions/monday/mutations/changeStatus.ts]: why don't server actions throw an error? should I remove the error logging entirely?

Comment (jinkang-0) [reply actions/monday/mutations/changeStatus.ts]: Servers (most often) operate on a separate machine from the client. Server actions get called when the client (i.e. the browser) sends a request to the server (i.e. some machine operated under Vercel, our hosting platform).

What happens when a server action crashes? Vercel probably runs the hosting service in a sandbox and catches the error, so realistically it may not crash, but the client doesn't know that. From the client's perspective, it sent a request, and it got back nothing.

In the code, we can anticipate errors. As you are doing with try-catch statements, once we catch an error, we can figure out what to do with it. Instead of throwing this error out after logging, we should tell the client that an error occurred. This provides information to the client about what went wrong.

You can reference some other server actions under the `/actions` folder. A common paradigm we've been using is using the `{ data, error }` object. When the action is successful, we fill the `data` field with the data, and set `error` to null. When the action is unsuccessful (e.g. errors), we set the `data` field to null and describe the error message in a human-readable format. For example: `{ data: null, error: "An unexpected error occurred." }`.

This is best practice, also echoed in [Next.js guide on server error handling](https://nextjs.org/learn/dashboard-app/error-handling):
> When going to production, you want to more gracefully show a message to the user when something unexpected happens.

The guide goes on to talk about an error page, but that's more applicable for frontend development. What we are doing is solely backend-backend communication, so we can omit this.

File: actions/monday/mutations/changeStatus.ts (hunk 1 above)
Comment (jinkang-0) [reply actions/monday/mutations/changeStatus.ts]: Simple: the function definition tells us it will throw an error
image

File: actions/monday/mutations/changeStatus.ts (hunk 3 above)
Comment (niranjana2222) [reply actions/monday/mutations/changeStatus.ts]: I see thanks!

File: actions/monday/mutations/changeStatus.ts (hunk 1 above)
Comment (niranjana2222) [reply actions/monday/mutations/changeStatus.ts]: Sounds good, also moved the MONDAY BOARD ID only to changeStatus.ts

**Comment by niranjana2222** [comment]
`import type { Database } from '@/types/database.types';
import { createClient } from '@supabase/supabase-js';
import { expect, test } from 'vitest';
import { mondayApiClient } from '@/actions/monday/core';
import updateAdopteeMondayStatus from '@/actions/monday/mutations/changeStatus';
import { loadEnvironment } from '@/lib/test-utils';

test('updateAdopteeMondayStatus handles OFC and both WL branches', async () => {
    loadEnvironment();
    const url = process.env.NEXT_PUBLIC_SUPABASE_URL;
    const key = process.env.SUPABASE_SERVICE_KEY;
    const boardId = process.env.MONDAY_WL_PIPS_BOARD_ID;
    const id = process.env.TEST_ADOPTEE_ID;
    if (!url || !key || !boardId || !id) {
      throw new Error(
        'Missing Supabase URL/key, MONDAY_WL_PIPS_BOARD_ID, or TEST_ADOPTEE_ID',
      );
    }

    const supabase = createClient<Database>(url, key);

    // stores past ids
    const statusRes = (await mondayApiClient.request(
      `query { items(ids: ["${id}"]) { column_values(ids: ["status__1"]) { text } } }`,
    )) as { items: { column_values: { text: string | null }[] }[] };
    const savedStatus = statusRes.items[0].column_values[0].text ?? '';
    const { data: faRow } = await supabase
      .from('adoptee_vector_test')
      .select('formerly_adopted')
      .eq('id', id)
      .single();
    const savedFormerlyAdopted = faRow!.formerly_adopted;

    // check ofc
    const ofcQuery = await updateAdopteeMondayStatus([id], 'OFC');
    expect(ofcQuery).toContain('OFC: Out For Consideration');

    // check wl
    await supabase
      .from('adoptee_vector_test')
      .update({ formerly_adopted: false })
      .eq('id', id);
    const wlQuery = await updateAdopteeMondayStatus([id], 'WL');
    expect(wlQuery).toContain('WL: Wait Listed');
    expect(wlQuery).not.toContain('WLFA');

    // check wlfa
    await supabase
      .from('adoptee_vector_test')
      .update({ formerly_adopted: true })
      .eq('id', id);
    const wlfaQuery = await updateAdopteeMondayStatus([id], 'WL');
    expect(wlfaQuery).toContain('WLFA: Wait Listed Formerly Adopted');

    // restore original
    await mondayApiClient.request(
      `mutation { change_simple_column_value(board_id: "${boardId}", item_id: "${id}", column_id: "status__1", value: ${JSON.stringify(savedStatus)}) { id } }`,
    );
    await supabase
      .from('adoptee_vector_test')
      .update({ formerly_adopted: savedFormerlyAdopted })
      .eq('id', id);
});
`

**Review by InayaY3 — APPROVED** [review]
great work! small note but in the future, please test before making a PR/claiming a PR is complete! if you're unsure about how to test, especially with unfamiliar API, don't be afraid to ask :)

fixed a small bug regarding monday status names, added a usable file for testing both WL/OFC changes, and restored some comments in `exportApplication`. also renamed some things for clarity. think that's all the major changes i made? definitely review to see the fixes. 

ty to ethan for fixing the rebase + prettier. again, those should be quick 5-10 min fixes! really slows review down if we have to fix everything for you

**future note: found while testing other PRs as well, but seems like monday automations don't always move rows back to the correct group, will mention in documentation
